update Access-Control-Allow-Headers to enumerate rather than wildcard

starpit opened the pull request:
As summarized in #4045, the previous #4010 PR, which changed Access-Control-Allow-Headers from a 2-element list to a wildcard was misguided. Despite whatwg introducing wildcard support in 2016, only Chrome as of yet supports this feature. So, for now, we need to back off to a specifically enumerated list of allowed headers. This PR reverts the wildcard change, but expands the list from the initial 2 to include a few more common headers, e.g. User-Agent.

Fixes #4045


<!--- Provide a concise summary of your changes in the Title -->

## Description
<!--- Provide a detailed description of your changes. -->
<!--- Include details of what problem you are solving and how your changes are tested. -->

## Related issue and scope
<!--- Please include a link to a related issue if there is one. -->
- [x] I opened an issue to propose and discuss this change (#4045)

## My changes affect the following components
<!--- Select below all system components are affected by your change. -->
<!--- Enter an `x` in all applicable boxes. -->
- [ ] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [x] Documentation

## Types of changes
<!--- What types of changes does your code introduce? Use `x` in all the boxes that apply: -->
- [x] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
<!--- Please review the points below which help you make sure you've covered all aspects of the change you're making. -->

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala:
```diff
@@ -151,7 +151,13 @@ protected[controller] object RestApiCommons {
+    "X-Requested-With",
+    "Content-Type",
+    "Accept",
+    "User-Agent")
```
Should we make this list configurable?

starpit replied:
probably the ultimate right course of action is to make both fields (allow-origin, allow-headers) configurable.

the reason i've put in this short-term fix is that my previous PR probably broke some use cases; whereas previously web actions would work from any browser (assuming only those two previously allowed headers were passed), now web actions will only work from chrome (but will support *any* headers)

with this PR, we restore functionality on all browsers for web actions, with a somewhat expanded list of allow headers)

i would suggest that we do the right thing in a subsequent PR, but this PR i feel fills an important a short-term regression fix until we can get the right PR in place?

markusthoemmes replied:
SGTM

markusthoemmes replied:
Can we at least define the list of headers in a centralized place so we don't end up copying them around?

starpit replied:
sounds good. i'll work on this in a bit.

starpit commented:
i have updated the PR to introduce a new file CorsSettings, and cleaned up RestAPIs and WebActions to pull the constants from this file.

in a subsequent PR, perhaps we can allow CorsSettings to be populated from the ansible/whatever level of deploy-time configuration.

starpit commented:
further refined to use traits

starpit commented:
ping
